File: engine/engine_client.py
"""Minimal Python client for the Berkeley engine."""

from __future__ import annotations

import json
from typing import Any
from urllib import request


class EngineClient:
    def __init__(self, base_url: str, client_id: str | None = None, client_token: str | None = None) -> None:
        self.base_url = base_url.rstrip("/")
        self.client_id = client_id
        self.client_token = client_token

    def ingest_message(self, payload: dict[str, Any]) -> dict[str, Any]:
        headers = {}
        if self.client_id and self.client_token:
            headers["X-CLIENT-ID"] = self.client_id
            headers["X-CLIENT-TOKEN"] = self.client_token
        return self._request("POST", "/api/ingest/message", payload, headers)

    def ask(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self._request("POST", "/api/ask", payload, {})

    def _request(self, method: str, path: str, payload: dict[str, Any], headers: dict[str, str]) -> dict[str, Any]:
        url = f"{self.base_url}{path}"
        data = json.dumps(payload).encode("utf-8")
        req = request.Request(url, data=data, method=method)
        req.add_header("Content-Type", "application/json")
        for key, value in headers.items():
            req.add_header(key, value)
        with request.urlopen(req) as resp:
            return json.loads(resp.read().decode("utf-8"))

File: engine/engine_client.ts
export type AskPayload = {
  query: string;
  scope?: Record<string, unknown>;
  mode?: "auto" | "knowledge" | "experience";
  viewer_role?: "student" | "admin";
  context?: Record<string, unknown>;
};

export class EngineClient {
  private baseUrl: string;
  private clientId?: string;
  private clientToken?: string;

  constructor(baseUrl: string, clientId?: string, clientToken?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.clientId = clientId;
    this.clientToken = clientToken;
  }

  async ingestMessage(payload: Record<string, unknown>): Promise<Record<string, unknown>> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.clientId && this.clientToken) {
      headers["X-CLIENT-ID"] = this.clientId;
      headers["X-CLIENT-TOKEN"] = this.clientToken;
    }
    const response = await fetch(`${this.baseUrl}/api/ingest/message`, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
    });
    return response.json();
  }

  async ask(payload: AskPayload): Promise<Record<string, unknown>> {
    const response = await fetch(`${this.baseUrl}/api/ask`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    return response.json();
  }
}
